File: sdk/javascript/tusk/protection.js
/**
 * TuskLang SDK Protection Core Module
 * Enterprise-grade protection for JavaScript SDK
 */

const crypto = require('crypto');

class TuskProtection {
    constructor(licenseKey, apiKey) {
        this.licenseKey = licenseKey;
        this.apiKey = apiKey;
        this.sessionId = this.generateUUID();
        this.integrityChecks = {};
        this.usageMetrics = {
            startTime: Date.now(),
            apiCalls: 0,
            errors: 0
        };
        this.encryptionKey = this.deriveKey(licenseKey);
    }

    generateUUID() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            const r = Math.random() * 16 | 0;
            const v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }

    deriveKey(password) {
        const salt = Buffer.from('tusklang_protection_salt', 'utf8');
        return crypto.pbkdf2Sync(password, salt, 100000, 32, 'sha256');
    }

    validateLicense() {
        try {
            if (!this.licenseKey || this.licenseKey.length < 32) {
                return false;
            }
            
            const checksum = crypto.createHash('sha256').update(this.licenseKey).digest('hex');
            return checksum.startsWith('tusk');
        } catch (error) {
            return false;
        }
    }

    encryptData(data) {
        try {
            const iv = crypto.randomBytes(16);
            const cipher = crypto.createCipher('aes-256-cbc', this.encryptionKey);
            let encrypted = cipher.update(data, 'utf8', 'hex');
            encrypted += cipher.final('hex');
            return iv.toString('hex') + ':' + encrypted;
        } catch (error) {
            return data;
        }
    }

    decryptData(encryptedData) {
        try {
            const parts = encryptedData.split(':');
            const iv = Buffer.from(parts[0], 'hex');
            const encrypted = parts[1];
            const decipher = crypto.createDecipher('aes-256-cbc', this.encryptionKey);
            let decrypted = decipher.update(encrypted, 'hex', 'utf8');
            decrypted += decipher.final('utf8');
            return decrypted;
        } catch (error) {
            return encryptedData;
        }
    }

    verifyIntegrity(data, signature) {
        try {
            const expectedSignature = crypto
                .createHmac('sha256', this.apiKey)
                .update(data)
                .digest('hex');
            return crypto.timingSafeEqual(
                Buffer.from(signature, 'hex'),
                Buffer.from(expectedSignature, 'hex')
            );
        } catch (error) {
            return false;
        }
    }

    generateSignature(data) {
        return crypto
            .createHmac('sha256', this.apiKey)
            .update(data)
            .digest('hex');
    }

    trackUsage(operation, success = true) {
        this.usageMetrics.apiCalls++;
        if (!success) {
            this.usageMetrics.errors++;
        }
    }

    getMetrics() {
        return {
            ...this.usageMetrics,
            sessionId: this.sessionId,
            uptime: Date.now() - this.usageMetrics.startTime
        };
    }

    obfuscateCode(code) {
        return Buffer.from(code).toString('base64');
    }

    detectTampering() {
        try {
            // Check module integrity
            const currentModule = __filename;
            const fs = require('fs');
            const content = fs.readFileSync(currentModule);
            const fileHash = crypto.createHash('sha256').update(content).digest('hex');
            this.integrityChecks[currentModule] = fileHash;
            return true;
        } catch (error) {
            return false;
        }
    }

    reportViolation(violationType, details) {
        const violation = {
            timestamp: Date.now(),
            sessionId: this.sessionId,
            type: violationType,
            details: details,
            licenseKey: this.licenseKey.substring(0, 8) + '...'
        };
        
        console.log(`SECURITY VIOLATION: ${JSON.stringify(violation)}`);
        return violation;
    }
}

// Global protection instance
let protectionInstance = null;

function initializeProtection(licenseKey, apiKey) {
    protectionInstance = new TuskProtection(licenseKey, apiKey);
    return protectionInstance;
}

function getProtection() {
    if (!protectionInstance) {
        throw new Error('Protection not initialized. Call initializeProtection() first.');
    }
    return protectionInstance;
}

module.exports = {
    TuskProtection,
    initializeProtection,
    getProtection
}; 

File: sdk/go/protection/protection.go
// Package protection provides TuskLang SDK Protection Core Module
// Enterprise-grade protection for Go SDK

package protection

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"hash"
	"sync"
	"time"

	"github.com/google/uuid"
	"golang.org/x/crypto/pbkdf2"
)

// UsageMetrics tracks SDK usage statistics
type UsageMetrics struct {
	StartTime int64 `json:"start_time"`
	APICalls  int64 `json:"api_calls"`
	Errors    int64 `json:"errors"`
}

// Violation represents a security violation
type Violation struct {
	Timestamp        int64  `json:"timestamp"`
	SessionID        string `json:"session_id"`
	ViolationType    string `json:"violation_type"`
	Details          string `json:"details"`
	LicenseKeyPartial string `json:"license_key_partial"`
}

// TuskProtection provides core protection functionality
type TuskProtection struct {
	licenseKey      string
	apiKey          string
	sessionID       string
	encryptionKey   []byte
	integrityChecks map[string]string
	usageMetrics    UsageMetrics
	mutex           sync.RWMutex
}

// New creates a new TuskProtection instance
func New(licenseKey, apiKey string) *TuskProtection {
	sessionID := uuid.New().String()
	encryptionKey := deriveKey(licenseKey)
	startTime := time.Now().Unix()

	return &TuskProtection{
		licenseKey:      licenseKey,
		apiKey:          apiKey,
		sessionID:       sessionID,
		encryptionKey:   encryptionKey,
		integrityChecks: make(map[string]string),
		usageMetrics: UsageMetrics{
			StartTime: startTime,
			APICalls:  0,
			Errors:    0,
		},
	}
}

func deriveKey(password string) []byte {
	salt := []byte("tusklang_protection_salt")
	return pbkdf2.Key([]byte(password), salt, 100000, 32, sha256.New)
}

// ValidateLicense checks if the license key is valid
func (tp *TuskProtection) ValidateLicense() bool {
	if len(tp.licenseKey) < 32 {
		return false
	}

	hash := sha256.Sum256([]byte(tp.licenseKey))
	checksum := hex.EncodeToString(hash[:])
	return len(checksum) >= 4 && checksum[:4] == "tusk"
}

// EncryptData encrypts sensitive data
func (tp *TuskProtection) EncryptData(data string) string {
	block, err := aes.NewCipher(tp.encryptionKey)
	if err != nil {
		return data
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return data
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := rand.Read(nonce); err != nil {
		return data
	}

	ciphertext := gcm.Seal(nonce, nonce, []byte(data), nil)
	return base64.StdEncoding.EncodeToString(ciphertext)
}

// DecryptData decrypts sensitive data
func (tp *TuskProtection) DecryptData(encryptedData string) string {
	ciphertext, err := base64.StdEncoding.DecodeString(encryptedData)
	if err != nil {
		return encryptedData
	}

	block, err := aes.NewCipher(tp.encryptionKey)
	if err != nil {
		return encryptedData
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return encryptedData
	}

	if len(ciphertext) < gcm.NonceSize() {
		return encryptedData
	}

	nonce, ciphertext := ciphertext[:gcm.NonceSize()], ciphertext[gcm.NonceSize():]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return encryptedData
	}

	return string(plaintext)
}

// VerifyIntegrity verifies data integrity using HMAC
func (tp *TuskProtection) VerifyIntegrity(data, signature string) bool {
	expectedSignature := tp.generateSignature(data)
	return hmac.Equal([]byte(signature), []byte(expectedSignature))
}

// GenerateSignature creates HMAC signature for data
func (tp *TuskProtection) GenerateSignature(data string) string {
	h := hmac.New(sha256.New, []byte(tp.apiKey))
	h.Write([]byte(data))
	return hex.EncodeToString(h.Sum(nil))
}

// TrackUsage records usage metrics
func (tp *TuskProtection) TrackUsage(operation string, success bool) {
	tp.mutex.Lock()
	defer tp.mutex.Unlock()

	tp.usageMetrics.APICalls++
	if !success {
		tp.usageMetrics.Errors++
	}
}

// GetMetrics returns usage metrics
func (tp *TuskProtection) GetMetrics() map[string]interface{} {
	tp.mutex.RLock()
	defer tp.mutex.RUnlock()

	uptime := time.Now().Unix() - tp.usageMetrics.StartTime
	return map[string]interface{}{
		"start_time":  tp.usageMetrics.StartTime,
		"api_calls":   tp.usageMetrics.APICalls,
		"errors":      tp.usageMetrics.Errors,
		"session_id":  tp.sessionID,
		"uptime":      uptime,
	}
}

// ObfuscateCode performs basic code obfuscation
func (tp *TuskProtection) ObfuscateCode(code string) string {
	return base64.StdEncoding.EncodeToString([]byte(code))
}

// DetectTampering checks if the SDK has been tampered with
func (tp *TuskProtection) DetectTampering() bool {
	// In production, implement file integrity checks
	// For now, return true as placeholder
	return true
}

// ReportViolation reports security violations
func (tp *TuskProtection) ReportViolation(violationType, details string) *Violation {
	violation := &Violation{
		Timestamp:        time.Now().Unix(),
		SessionID:        tp.sessionID,
		ViolationType:    violationType,
		Details:          details,
		LicenseKeyPartial: tp.licenseKey[:min(8, len(tp.licenseKey))] + "...",
	}

	fmt.Printf("SECURITY VIOLATION: %+v\n", violation)
	return violation
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// Global protection instance
var (
	protectionInstance *TuskProtection
	instanceMutex      sync.RWMutex
)

// InitializeProtection creates the global protection instance
func InitializeProtection(licenseKey, apiKey string) *TuskProtection {
	instanceMutex.Lock()
	defer instanceMutex.Unlock()

	protectionInstance = New(licenseKey, apiKey)
	return protectionInstance
}

// GetProtection returns the global protection instance
func GetProtection() *TuskProtection {
	instanceMutex.RLock()
	defer instanceMutex.RUnlock()

	if protectionInstance == nil {
		panic("Protection not initialized. Call InitializeProtection() first.")
	}
	return protectionInstance
} 